swagger doc for analysis ingest 404s (DSDEV-1866)

davidangb opened the pull request:
See my comment in DSDEV-1866 - changing the response code from 404 to 400 (or something else) is somewhat intrusive. So, I'm keeping this change small and limited to just the doc.

davidangb committed:
update swagger api doc for analysis ingest to reflect 404s for invalid input ids; expanded tests for analysis ingest.

jmthibault79 commented:
Looks good.

I would also like to see tests which capture the current behavior, including cases with >1 valid vault IDs and >1 vault IDs with a mix of valid and invalid.

davidangb commented:
Updated to include a bunch of extra tests for analysis-ingest, in varying combinations of valid and invalid input IDs.

Note that I took a different approach, using BeforeAndAfterAll, for creating the test data. Downside is that this is the only test in the repo that uses this approach. However, I like the approach better - it makes it clear which parts are setup and which parts are tests. If we fail during setup - i.e. inside beforeAll() - the entire test suite will abort, which is also nice. Plus it gives us a hook to clean up the test data later, once delete APIs exist.

See http://www.scalatest.org/user_guide/sharing_fixtures. I tried doing this with the "real" withFixture(OneArgTest) approach, but that was scala hell. I needed to change VaultFreeSpec to extend fixture.FreeSpec, which affected ALL tests. I tried letting AnalysisIngestServiceSpec extend VaultFreeSpec (which extends FreeSpec) and add "with fixture.FreeSpecLike" ... but having both FreeSpec and fixture.FreeSpecLike on the same class caused all kinds of conflicts. The approach I took seemed okay, even though it uses an instance var.

jmthibault79 commented on src/main/scala/org/broadinstitute/dsde/vault/services/analysis/AnalysisIngestService.scala:
```diff
@@ -21,13 +21,15 @@ trait AnalysisIngestService extends HttpService {
     notes = "Accepts a json packet as POST. Creates a Vault object with the supplied metadata and creates relationships for each input id. " +
-      " Returns the Vault ID of the created object.")
+      " Returns the Vault ID of the created object.\n\nThe values of the 'input' array must be valid Vault IDs for the ubams used as input to this analysis." +
+      " If an invalid id is specified inside the input array, this API will fail with a 404 response code.")
```
Does this need to be updated in accordance with #46 ?

jmthibault79 commented:
I like the new tests.  Not sure about the new fixture approach - it looks like a mix of the BeforeAndAfter and Loan patterns, and I'd prefer sticking to one.

davidangb committed:
use files that actually exist in GCS for tests, so boss won't complain

davidangb commented:
- updated swagger doc again to address the weird three-line error from #46. Good call.
- totally correct I was mixing the BeforeAndAfter and loan-fixture patterns. I removed loan-fixture and am using only BeforeAndAfter.
- unrelated to this PR but necessary to pass tests - I updated all the tests to reference files that actually exist in GCS (i.e. "vault/test/test.bam" instead of "/path/to/ingest/bam"). This is necessary because BOSS now validates existence of these files when using X-Force-Location; tests that previously passed now fail due to BOSS's more rigorous checking. I do not know why this suddenly became true; I assume boss-dev was somehow not up to date with the latest code.

ready for your re-review ... sorry this got bigger! I suppose I could have separated these.

jmthibault79 commented:
:+1: 

Please rebase and squash to 2 commits: BOSS filename updates, everything for DSDEV-1866.  Then you can merge it yourself.
